--- lib/libssl/.patch	1970-01-01 01:00:00 +0100
+++ lib/libssl/.patch	2025-09-12 11:34:13 +0200
@@ -0,0 +1 @@
+This directory has been patched.
--- lib/libssl/src/crypto/asn1/asn1.h	2025-10-19 08:07:00 +0200
+++ lib/libssl/src/crypto/asn1/asn1.h	2025-10-19 11:00:14 +0200
@@ -403,7 +403,7 @@
 #define ASN1_ITEM_rptr(ref) (&(ref##_it))
 
 #define DECLARE_ASN1_ITEM(name) \
-	OPENSSL_EXTERN const ASN1_ITEM name##_it;
+	OPENSSL_EXTERN __import const ASN1_ITEM name##_it;
 
 #else
 
