--- openbsd-master/usr.bin/.patch	1970-01-01 01:00:00 +0100
+++ openbsd-master/usr.bin/.patch	2025-09-12 11:34:13 +0200
@@ -0,0 +1 @@
+This directory has been patched.
--- -	Tue Oct  7 16:29:59 2025
+++ openbsd-master/usr.bin/ftp/Makefile	Sun Feb 16 19:02:22 2025
@@ -15,11 +15,12 @@
 SRCS=	cmds.c cmdtab.c complete.c cookie.c domacro.c fetch.c ftp.c \
 	list.c main.c ruserpass.c small.c stringlist.c util.c
 
+CFLAGS+=-Dsignal=bsd_signal
 CPPFLAGS+= -DINET6
 
 LDADD+=	-ledit -lcurses -lutil -lssl -lcrypto
 DPADD+=	${LIBEDIT} ${LIBCURSES} ${LIBUTIL}
-LDSTATIC= ${STATIC}
+#LDSTATIC= ${STATIC}
 
 #COPTS+= -Wall -Wconversion -Wstrict-prototypes -Wmissing-prototypes
 
--- -	Tue Oct  7 16:30:02 2025
+++ openbsd-master/usr.bin/sqlite3/Makefile	Sun May  1 08:20:56 2022
@@ -11,8 +11,8 @@
 
 CPPFLAGS +=	-I${.CURDIR}/../../lib/libedit -DHAVE_READLINE=1
 
-CFLAGS += -pthread
-LDFLAGS += -pthread
+#CFLAGS += -pthread
+#LDFLAGS += -pthread
 LDADD += -lsqlite3 -ledit -lcurses
 
 .include <bsd.prog.mk>
--- -	Tue Oct  7 16:30:02 2025
+++ openbsd-master/usr.bin/ssh/scp/Makefile	Sat Sep  6 07:38:24 2025
@@ -12,4 +12,6 @@
 
 SRCS=	scp.c
 
+LDADD+= -lssp
+
 .include <bsd.prog.mk>
--- -	Tue Oct  7 16:30:02 2025
+++ openbsd-master/usr.bin/ssh/sftp/Makefile	Sat Sep  6 07:41:46 2025
@@ -14,5 +14,5 @@
 
 .include <bsd.prog.mk>
 
-LDADD+=	-ledit -ltermcap -lutil
+LDADD+=	-ledit -ltermcap -lutil -lssp
 DPADD+= ${LIBEDIT} ${LIBTERMCAP} ${LIBUTIL}
--- -	Tue Oct  7 16:30:02 2025
+++ openbsd-master/usr.bin/ssh/sftp-server/Makefile	Sat Sep  6 07:39:16 2025
@@ -12,7 +12,7 @@
 
 SRCS=	sftp-server.c sftp-common.c sftp-server-main.c
 
-LDADD+=	-lutil
+LDADD+=	-lutil -lssp
 DPADD+= ${LIBUTIL}
 
 .include <bsd.prog.mk>
--- -	Tue Oct  7 16:30:02 2025
+++ openbsd-master/usr.bin/ssh/ssh/Makefile	Sat Sep  6 07:14:43 2025
@@ -42,5 +42,5 @@
 DPADD+=	${LIBCRYPTO} ${LIBUTIL}
 .endif
 
-LDADD+=	-lz
+LDADD+=	-lz -lssp
 DPADD+=	${LIBZ}
--- -	Tue Oct  7 16:30:03 2025
+++ openbsd-master/usr.bin/ssh/ssh-add/Makefile	Sat Sep  6 07:32:25 2025
@@ -14,5 +14,5 @@
 
 .include <bsd.prog.mk>
 
-LDADD+=	-lcrypto
+LDADD+=	-lcrypto -lssp
 DPADD+= ${LIBCRYPTO}
--- -	Tue Oct  7 16:30:03 2025
+++ openbsd-master/usr.bin/ssh/ssh-agent/Makefile	Sat Sep  6 07:36:51 2025
@@ -15,5 +15,5 @@
 
 .include <bsd.prog.mk>
 
-LDADD+=	-lcrypto
+LDADD+=	-lcrypto -lssp
 DPADD+=	${LIBCRYPTO}
--- -	Tue Oct  7 16:30:03 2025
+++ openbsd-master/usr.bin/ssh/ssh-keygen/Makefile	Sat Sep  6 07:33:46 2025
@@ -14,5 +14,5 @@
 
 .include <bsd.prog.mk>
 
-LDADD+=	-lcrypto
+LDADD+=	-lcrypto -lssp
 DPADD+=	${LIBCRYPTO}
--- -	Tue Oct  7 16:30:03 2025
+++ openbsd-master/usr.bin/ssh/ssh-keyscan/Makefile	Sat Sep  6 07:41:00 2025
@@ -14,5 +14,5 @@
 
 .include <bsd.prog.mk>
 
-LDADD+= -lcrypto -lz
+LDADD+= -lcrypto -lz -lssp
 DPADD+= ${LIBCRYPTO} ${LIBZ}
--- -	Tue Oct  7 16:30:03 2025
+++ openbsd-master/usr.bin/ssh/ssh-keysign/Makefile	Sat Sep  6 07:40:15 2025
@@ -14,5 +14,5 @@
 
 .include <bsd.prog.mk>
 
-LDADD+=	-lcrypto -lutil -lz
+LDADD+=	-lcrypto -lutil -lz -lssp
 DPADD+=	${LIBCRYPTO} ${LIBUTIL} ${LIBZ}
--- -	Tue Oct  7 16:30:03 2025
+++ openbsd-master/usr.bin/ssh/ssh-pkcs11-helper/Makefile	Sat Sep  6 07:42:27 2025
@@ -14,5 +14,5 @@
 
 .include <bsd.prog.mk>
 
-LDADD+=	-lcrypto
+LDADD+=	-lcrypto -lssp
 DPADD+=	${LIBCRYPTO}
--- -	Tue Oct  7 16:30:03 2025
+++ openbsd-master/usr.bin/ssh/sshd/Makefile	Tue Sep 23 19:46:23 2025
@@ -47,5 +47,5 @@
 DPADD+=	${LIBCRYPTO} ${LIBUTIL}
 .endif
 
-LDADD+=	-lz
+LDADD+=	-lz -lssp
 DPADD+=	${LIBZ}
--- -	Tue Oct  7 16:30:05 2025
+++ openbsd-master/usr.bin/top/Makefile	Wed Sep 17 05:02:47 2025
@@ -4,7 +4,7 @@
 
 PROG=	top
 
-CFLAGS+=-I. -Wall -Wmissing-prototypes
+CFLAGS+=-I. -Wall -Wmissing-prototypes -Dsignal=bsd_signal
 SRCS=	commands.c display.c machine.c screen.c top.c username.c utils.c
 DPADD=	${LIBCURSES}
 LDADD=	-lcurses
--- -	Tue Oct  7 16:30:05 2025
+++ openbsd-master/usr.bin/vi/build/config.h	Tue Jul  8 07:59:03 2025
@@ -102,7 +102,7 @@
 /* #undef HAVE_LOCK_FCNTL */
 
 /* Define if you have flock(2) style locking. */
-#define HAVE_LOCK_FLOCK 1
+/* #define HAVE_LOCK_FLOCK 1 */
 
 /* Define if you want to compile in the Perl interpreter. */
 /* #undef HAVE_PERL_INTERP */
--- -	Tue Oct  7 16:30:06 2025
+++ openbsd-master/usr.bin/xinstall/Makefile	Thu Jun 26 10:58:26 2025
@@ -5,7 +5,7 @@
 MAN=	install.1
 
 realinstall:
-	${INSTALL} ${INSTALL_COPY} -S ${INSTALL_STRIP} \
+	./xinstall ${INSTALL_COPY} -S ${INSTALL_STRIP} \
 	    -o ${BINOWN} -g ${BINGRP} \
 	    -m ${BINMODE} ${PROG} ${DESTDIR}${BINDIR}/install
 
--- -	Tue Oct  7 16:30:06 2025
+++ openbsd-master/usr.bin/xinstall/xinstall.c	Sun Jul 20 06:53:18 2025
@@ -272,13 +272,13 @@
 	}
 
 	if (dostrip) {
+		close(to_fd);
 		strip(safecopy ? tempfile : to_name);
 
 		/*
 		 * Re-open our fd on the target, in case we used a strip
 		 *  that does not work in-place -- like gnu binutils strip.
 		 */
-		close(to_fd);
 		if ((to_fd = open(safecopy ? tempfile : to_name, O_RDONLY,
 		     0)) < 0)
 			err(EX_OSERR, "stripping %s", to_name);
